services/venice-server/src/main/java/com/linkedin/venice/listener/ServerReadMetadataRepository.java

```diff
@@ -72,6 +72,11 @@ public MetadataResponse getMetadata(String storeName) {
                 "Fast client is not enabled for store: %s, please ensure storage node read quota is enabled for the given store",
                 storeName));
       }
+      if (store.isMigrating()) {
```

Comment:
This check is incomplete. We only throw exception if the `store.isMigrating()` AND the cluster has changed to destination cluster. i.e. complete migration is called but we haven't ended migration yet. This is the window we should be redirecting. Too early won't do the trick.

Reply:
@xunyin8 Something occurred to me about this logic. Say, at the stage 2, current cluster name is same as dest cluster so a refresh happens but if the `end migration` command is not sent timely, the the client actually would keep getting exceptions because isMigrating flag will only be set to false after `end migration` command is received.

`end migration` can only be called when all regions has completed migration which means if any of regions is lagging behind, the clients in other regions couldn't proceed due to this check.

Reply:
If the server tells clients which cluster it should belongs to and client has a way to verify that (e.g. using `serverD2ServiceName` to infer the cluster name so it can choose to refresh or not.  WDYT?

Reply:
Ok, let's take a step back and outline the exact conditions we are checking and we can go over the different cases. We can even include the case where the migration was completed but we decided to abort last second for whatever reason.

The conditions we are checking and will be throwing exception are (both needs to be true):
1. Store is migrating
2. Cluster info does not match with the cluster that the server is responding to the request. e.g. store cluster config is v-1 but the server responding is in v-0.

In the case you described, once we re-discover and start hitting server in v-1 we will no longer throw exceptions and metadata request will go through and we will be getting routing table in v-1 and traffic will be going to v-1.

It'll be tricky if we decided to abort migration after we completed migration for whatever reason. Store is no longer migrating, we reset the cluster info and deleted the store in dest cluster. This will cause an issue and I feel like we need to improve this even for other clients (TC) as well. We should also split abort into two steps like complete+end. We should only delete stores in dest or src in end and both abort and complete should just be changing the cluster info. If we do that then we will also get a window to redirect traffic back to src cluster.

Reply:
I see why we have this misunderstanding. You are saying we check the server cluster name with the cluster set for a given store from `StoreConfig`. I thought you suggest we compare the dest cluster name with the cluster set for a given store.  The former makes more sense.

Reply:
For the abort-migration after complete-migration, if it is tricky to support it, maybe we should just disallow such action as the operator can always finish the migration and migrate it back to the original cluster if there is a need, but I don't think this happen often.

Reply:
I checked the code and abort migration is a two step approach already unless we provided with override flag. So I think we are good in that front too. @gaojieliu Why is it tricky? Today we just rever the cluster discovery back to original and can stay in that state without deleting the cloned store until traffic is fully shifted back.